src/Lorentz/Contracts/BaseDAO/FA2.hs

```diff
@@ -0,0 +1,185 @@
+  (https://gitlab.com/tzip/tzip/-/blob/667f925471728bb8e81fbd30b93a171e401b6dc1/proposals/tzip-12/tzip-12.md)
+  |]
+
+
+fa2LorentzContract :: Contract Parameter Storage
+fa2LorentzContract = defaultContract $ contractName "FA2 smart contract" $ do
+  doc $ DDescription contractDoc
+  unpair
+  entryCase @Parameter (Proxy @PlainEntrypointsKind)
+    ( #cTransfer /-> transfer
+    , #cBalance_of /-> balanceOf
+    , #cToken_metadata_registry /-> tokenMetadataRegistry
+    , #cPermissions_descriptor /-> permissionsDescriptor
+    , #cUpdate_operators /-> updateOperators
+    )
+
+-- | TODO: Implement missing methods
+transfer :: Entrypoint TransferParams Storage
+transfer = doNothing
+
+balanceOf :: Entrypoint BalanceRequestParams Storage
+balanceOf = doNothing
+
+permissionsDescriptor :: Entrypoint PermissionsDescriptorParam Storage
+permissionsDescriptor = doNothing
+
+doNothing :: Entrypoint a b
+doNothing = do
+  drop
+  nil
+  pair
+
+tokenMetadataRegistry :: Entrypoint TokenMetadataRegistryParam Storage
+tokenMetadataRegistry = do
+  swap
+  getField #sTokenAddress
+  dig @2
+  swap
+  push @Mutez (toMutez 0)
+  swap
+  transferTokens # nil # swap # cons # pair
+
+updateOperators :: Entrypoint UpdateOperatorsParam Storage
+updateOperators = do
+  iter $
+    caseT @UpdateOperator
+      ( #cAdd_operator /-> addOperator
+      , #cRemove_operator /-> removeOperator
+      )
+  nil; pair
+
+addOperator :: forall s. OperatorParam & Storage & s :-> Storage & s
+addOperator = do
+  unpair
+  fromNamed #owner
+  swap
+  unpair
+  fromNamed #operator
+  swap
+  fromNamed #token_id
+  dup
+  assertNeq0or1
+  dug @2
+  swap
+  dupTop2
+  assertNeq [mt|NOT_OWNER|]
+  dug @2
+  pair
+  swap
+  pair
+  swap
+  getField #sOperators
+  dig @2
+  dupTop2
+  mem
+  if_ ifKeyExists ifKeyDoesntExist
+  where
+    ifKeyExists :: (Address, (Address, TokenId)) & Operators & Storage & s :-> Storage & s
+    ifKeyExists = dropN @2
+
+    ifKeyDoesntExist :: (Address, (Address, TokenId)) & Operators & Storage & s :-> Storage & s
+    ifKeyDoesntExist = do
+      unit
+      some
+      swap
+      update
+      setField #sOperators
+
+removeOperator :: forall s. OperatorParam & Storage & s :-> Storage & s
+removeOperator = do
+  unpair
+  fromNamed #owner
+  swap
+  unpair
+  fromNamed #operator
+  swap
+  fromNamed #token_id
+  dup
+  assertNeq0or1
+  dug @2
+  swap
+  dupTop2
+  assertNeq [mt|NOT_OWNER|]
+  dug @2
+  pair
+  swap
+  pair
+  swap
+  getField #sOperators
+  dig @2
+  dupTop2
+  mem
+  if_ ifKeyExists ifKeyDoesntExist
+  where
+    ifKeyExists :: (Address, (Address, TokenId)) & Operators & Storage & s :-> Storage & s
+    ifKeyExists = do
+      push Nothing
+      swap
+      update
+      setField #sOperators
+
+
+    ifKeyDoesntExist :: (Address, (Address, TokenId)) & Operators & Storage & s :-> Storage & s
+    ifKeyDoesntExist = dropN @2
+
+-- | TODO: Move this one to Lorentz
```

Comment:
Btw, you are welcome to review https://gitlab.com/morley-framework/morley/-/merge_requests/591. Let's keep it here for now, we will include it into the next `lorentz` release and update this place.